plugins/inputs/intel_dlb/intel_dlb.go

```diff
@@ -0,0 +1,470 @@
+//go:generate ../../../tools/readme_config_includer/generator
+//go:build linux
+// +build linux
+
+package intel_dlb
+
+import (
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal/choice"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+type IntelDLB struct {
+	SocketPath       string          `toml:"socket_path"`
+	EventdevCommands []string        `toml:"eventdev_commands"`
+	DLBDeviceIDs     []string        `toml:"dlb_device_types"`
+	Log              telegraf.Logger `toml:"-"`
+
+	connection           net.Conn
+	devicesDir           []string
+	rasReader            rasReader
+	maxInitMessageLength uint32
+	logOnce              map[string]error
+}
+
+// DO NOT REMOVE THE NEXT TWO LINES! This is required to embed the sampleConfig data.
+//
+//go:embed sample.conf
+var sampleConfig string
+
+const (
+	defaultSocketPath      = "/var/run/dpdk/rte/dpdk_telemetry.v2"
+	pluginName             = "intel_dlb"
+	eventdevListCommand    = "/eventdev/dev_list"
+	dlbDeviceIDLocation    = "/sys/devices/*/*/device"
+	aerCorrectableFileName = "aer_dev_correctable"
+	aerFatalFileName       = "aer_dev_fatal"
+	aerNonFatalFileName    = "aer_dev_nonfatal"
+	defaultDLBDevice       = "0x2710"
+)
+
+// SampleConfig returns sample config
+func (d *IntelDLB) SampleConfig() string {
+	return sampleConfig
+}
+
+// Init performs validation of all parameters from configuration.
+func (d *IntelDLB) Init() error {
+	var err error
+
+	if d.SocketPath == "" {
+		d.SocketPath = defaultSocketPath
+		d.Log.Debugf("Using default '%v' path for socket_path", defaultSocketPath)
+	}
+
+	err = checkSocketPath(d.SocketPath)
+	d.logErrorOnce("socketPathDoesNotExist", err)
```

Comment:
Should we return the error at this point?

Reply:
can you explain later? 

This feels like something we should absolutely error out on unless the user specifically says ignore start up errors via a configuration option.

Reply:
@powersj 
It is just following guidelines from https://github.com/influxdata/telegraf/blob/master/docs/developers/REVIEWS.md#reviewing-plugin-code :

"`Init() error` should not contain connections to external services.`"

The same situation happens for example in [http plugin](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/http/http.go#L53) which doesn't check in `Init()` if all URLs are available during startup or in [file plug](https://github.com/influxdata/telegraf/blob/master/plugins/inputs/file/file.go#L36) which doesn't check in `Init()` if all files are available during startup.

User is informed about absence of socket once (in log file), so I think that we don't need to overcomplicate situation by introducing new configuration option.

Reply:
Without that socket a user cannot grab metrics, correct?

Reply:
> Without that socket a user cannot grab metrics, correct?

This statement is valid for file or tail plugin which will not be able to parse anything if observed files don't exist YET :)

For `intel_dlb` it is partially valid because socket (which can be created after Telegraf starts) is not single source of metrics. They are also grabbed from `sysfs`.

Reply:
Then in gather this needs to be checked and the plugin needs to error out if the socket is not available. This is how other plugins that look at sockets or remote services work.

If you want to have different behavior that ignores errors, then like what was recently added to mongodb, it needs to be put behind a config option.

Reply:
I think we should talk about this in the maintainers meeting :) because my understanding is if a socket/service/device that an input is monitoring is not available we should be erroring out -> telegraf should not start. These situations need to be clear calls to action to the end-user that something is not write and telegraf is not collecting metrics. 

We get bugs about this behavior with re: kafka, mongo, etc. The answer being get your infrastructure in place, make sure you don't have any typos in your config, and now, if available, use a config option to ignore these errors.

With respect to your question about partial metrics, that is new to me. But in this specific case, if the socket is not available it seems that we should not allow Telegraf to continue unless some additional config option is set to ignore this. I don't believe

Reply:
Why only socket/service/device? What about files/URIs and so on?
All of these "external services" may be unavailable during Telegraf start due to a lot of different reasons: Internet is down (because Telegraf service started before connection to router was established after OS had started), device has not been mounted yet, file is not available yet (maybe it is on CIFS which is not yet available) etc...

That's why we can read here https://github.com/influxdata/telegraf/blob/fb7a886d1882ac9689da757a1ab291c8e91be40b/docs/developers/REVIEWS.md?plain=1#L43

In above examples it is not a configuration error, something is just not yet available. And it may be available in the nearest future. `err` in `Init()` should be only returned if there is a configuration error (plugin configured in that way, that it is certain that no metrics can be exposed without configuration change, forbidden values in configuration fields and so on).

Of course I understand that end-user can make typo in plugin config. But for situation when it cannot be distinguished if connection/read failed because of typo or because something is not yet available Telegraf should not fail to start. User should be informed about the situation but Telegraf should work normally.

You mentioned mongodb plugin and putting some behavior behind some options... If this plugin had been written correctly from the beginning (so connecting to services in `Gather()` instead of in `Init()`) no bugs would have been submitted in this topic. Moreover, last PR which introduced this config in mongodb plugin introduced also some config checking in `Start()` function which stands in opposition to another guideline:
https://github.com/influxdata/telegraf/blob/fb7a886d1882ac9689da757a1ab291c8e91be40b/docs/developers/REVIEWS.md?plain=1#L42


> my understanding is if a socket/service/device that an input is monitoring is not available we should be erroring out -> telegraf should not start

Ok, so what would be your understanding about following configuration of tail plugin:
```
[[inputs.tail]]
  files = ["/tmpy/mylog.txt", "/mnt/floppy/very_old.log"]
  from_beginning = true
```
- `/tmpy/mylog.txt` is not available when Telegraf starts. Probably it is a typo, there is no `/tmpy` directory, user meant `tmp` right? Or maybe it is valid path but service which logs there, will start in 10 seconds?
- `/mnt/floppy/very_old.log` there is some log to read from my floppy disc which hasn't been mounted yet. But will it? Maybe my floppy drive is dead 😢 Should Telegraf spam about missing path to this monitored input every 10 seconds?

I'm very curious about your understanding about this scenario. What should happen here? Should Telegraf fail to start? If not, what logs should appear?

Reply:
I'm not sure it's a good idea to look at other plugins as specific examples of when an error condition should stop telegraf from running and when it should be logged. There are many plugins and enough exceptions to guidelines across plugins that it's likely we could find an example for whatever behavior we want :smile: 

The guidelines come down to 1) it's best for telegraf not to hide big errors but 2) it should do its best to function.

Generally if a plugin has a filename setting in the local filesystem but that file doesn't exist, telegraf should fail to start. Otherwise typos or other mistakes will cause unexpected behavior without warning.

If a plugin has a URL setting for a resource on the network, for example if it's fetching a web page each gather interval, the plugin should warn if the resource is temporarily unavailable but it shouldn't prevent telegraf from running.

The difficulty comes when plugins are expected to come up in an error state, like your example of unmounted filesystem. There's no way to know whether a filesystem will ever be mounted and bring telegraf out of error state, or if it's a typo in the setting that will never be fixed. In these cases I think telegraf should default to assuming the worst, so it should print an error and fail to start. If there is a clear use case that requires telegraf to start before a resource is available, maybe there should be an opt-in setting to allow that.

For this plugin's socket setting, since it's not a network resource, but a unix socket that must exist in the local filesystem for the plugin to be able to do its job, telegraf should expect it to exist and to fail to start if it doesn't exist. As far as I can tell this isn't a case where telegraf should be expected to start before the resource is available.

Reply:
@reimda This unix socket is exposed by application which may or may not be running during Telegraf start. This application may be stopped or restarted. And it is perfectly normal behaviour.

Lets assume that we follow your way, so please consider these two simple scenarios:
1. Socket is not available when Telegraf starts. Telegraf fail to start. Socket become available 5 seconds later.
2. Socket is available when Telegraf starts. After 5 seconds application (which exposes this socket) stops for some reason and socket is not available anymore. Telegraf had started successfully but will fail to `Gather` metrics from this socket.

It is all about timing (difference: 5 seconds) but results are totally different. It seems that there is lack of consistency in your approach.

2nd thing is (as I wrote before) that socket is only one source of metrics in this plugin. `sysfs` is the second one. Plugin can work perfectly fine exposing only part of metrics, hoping that in the near future socket will become available and all metrics can be exposed :)

Reply:
According to our agreement, plugin has been changed and will fail by default in `Init()` when socket is not available during this `Init()`.

Additional configuration field(`unreachable_socket_behavior`) was added to configuration to override this behavior (for more advanced users). With this option (`unreachable_socket_behavior = "ignore"`) plugin will ignore socket not being available during `Init()`. This was done for situations (described in this thread) when it s normal that socket is not available during `Init()` but will become available after some time.